apps/builder/app/builder/features/ai/hooks/media-recorder.ts

```diff
@@ -0,0 +1,159 @@
+import { useRef, useState } from "react";
+import { useEffectEvent } from "./effect-event";
+
+// https://github.com/openai/whisper/discussions/870
+const DEFAULT_OPTIONS: MediaRecorderOptions = {
+  audioBitsPerSecond: 16000,
+};
+
+export const useMediaRecorder = (
+  {
+    onComplete,
+    onReportSoundAmplitude,
+  }: {
+    onComplete: (file: File) => void;
+    onReportSoundAmplitude?: (amplitude: number) => void;
+  },
+  options = DEFAULT_OPTIONS
+) => {
+  const disposeRef = useRef<() => void>();
+
+  const cancelRef = useRef(false);
+  const isActiveRef = useRef<boolean>(false);
+  const idRef = useRef(0);
+  const [error, setError] = useState<Error>();
+  const [state, setState] = useState<"inactive" | "recording">("inactive");
+
+  const start = useEffectEvent(async () => {
+    isActiveRef.current = true;
+    const chunks: Blob[] = [];
+    idRef.current++;
+    const id = idRef.current;
+
+    cancelRef.current = false;
+
+    let stream: MediaStream;
+    try {
+      stream = await navigator.mediaDevices.getUserMedia({
+        audio: true,
+        video: false,
+      });
+    } catch (error) {
+      // Not allowed, do not start new recording
+      isActiveRef.current = false;
+
+      if (error instanceof Error) {
+        setError(error);
+        return;
+      }
+      setError(new Error("Unknown error"));
+      return;
+    }
+
+    // Recording was stopped, do not start new recording
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore wrong ts error based on isActiveRef.current = true above it think that isActiveRef.current is always true
+    if (isActiveRef.current === false) {
+      stream.getAudioTracks().forEach((track) => track.stop());
+      return;
+    }
+
+    // New recording started, do cleanup and return
+    if (id !== idRef.current) {
+      stream.getAudioTracks().forEach((track) => track.stop());
+      return;
+    }
+
+    setError(undefined);
+    setState("recording");
+
+    const extension = MediaRecorder.isTypeSupported("audio/webm; codecs=opus")
```

Comment:
nit:
```suggestion
    const format = MediaRecorder.isTypeSupported("audio/webm; codecs=opus")
```

Reply:
subtype is the right name, https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#structure_of_a_mime_type
renamed